File: tests/unit/test_prm.cpp
// Roadmap R5 verify — PrmPlanner (multi-query roadmap planner):
//   • builds a roadmap over the static scene (nodes + edges), reported in PrmBuildStats;
//   • solves a start→goal that must route around a wall; the path is independently collision-free;
//   • ONE roadmap answers MANY queries (the multi-query point);
//   • determinism per (build seed, query seed); loud failure on the registry stub + null args.
#include <gtest/gtest.h>

#include <algorithm>
#include <memory>
#include <vector>

#include <Eigen/Core>

#include "quevedomp/collision/collision_scene.hpp"
#include "quevedomp/collision/edge_check.hpp"
#include "quevedomp/collision/geometry.hpp"
#include "quevedomp/planning/planner.hpp"
#include "quevedomp/planning/roadmap.hpp"
#include "quevedomp/planning/types.hpp"
#include "quevedomp/robot/robot_instance.hpp"
#include "quevedomp/robot/robot_model.hpp"

using namespace quevedomp;
using namespace quevedomp::planning;

namespace {

const char *kGantry2D = R"(<robot name="gantry2d">
  <link name="base"/>
  <joint name="jx" type="prismatic">
    <parent link="base"/><child link="cx"/>
    <origin xyz="0 0 0"/><axis xyz="1 0 0"/>
    <limit lower="-2" upper="2" effort="10" velocity="1"/>
  </joint>
  <link name="cx"/>
  <joint name="jy" type="prismatic">
    <parent link="cx"/><child link="ee"/>
    <origin xyz="0 0 0"/><axis xyz="0 1 0"/>
    <limit lower="-2" upper="2" effort="10" velocity="1"/>
  </joint>
  <link name="ee">
    <collision><geometry><sphere radius="0.1"/></geometry></collision>
  </link>
</robot>)";

JointPosition q2(double x, double y) {
  JointPosition q(2);
  q << x, y;
  return q;
}

struct Fixture {
  std::shared_ptr<const RobotModel> model;
  std::shared_ptr<const RobotInstance> robot;
  std::shared_ptr<collision::CollisionScene> scene;
};

Fixture make_fixture(bool with_wall) {
  Fixture f;
  f.model = RobotModel::from_urdf(kGantry2D);
  f.robot = std::make_shared<const RobotInstance>(f.model);
  collision::SceneDescription env;
  if (with_wall) {
    collision::BoxShape wall;
    wall.half_extents = Eigen::Vector3d(0.1, 1.25, 0.5); // padded x∈[-0.2,0.2], y up to ~0.6
    env.objects.push_back(
        {"wall", wall, Transform::from_translation(Eigen::Vector3d(0, -0.75, 0))});
  }
  f.scene = make_static_scene(f.model, env);
  return f;
}

PrmParams params(std::uint64_t seed) {
  PrmParams p;
  p.num_nodes = 800;
  p.k_neighbors = 12;
  // Edges (and the smoother) validate at this step; the independent path_is_free checks below use
  // the SAME resolution, so the collision guarantee and the verification are consistent (a finer
  // check would flag grazes the planner never promised to avoid — the discretization gap).
  p.edge_resolution = 0.02;
  p.seed = seed;
  return p;
}

bool path_is_free(const Fixture &f, const Path &path, double res) {
  const auto ws = f.scene->make_workspace();
  collision::QueryOptions opts;
  for (std::size_t i = 0; i + 1 < path.size(); ++i) {
    const auto e = collision::check_edge(*f.scene, *f.robot, path[i], path[i + 1],
                                         static_cast<float>(res), opts, *ws);
    if (!e.valid) {
      return false;
    }
  }
  return true;
}

PlanningProblem problem(const JointPosition &start, const JointPosition &goal, std::uint64_t seed) {
  PlanningProblem prob;
  prob.start = start;
  prob.goal = std::make_shared<JointGoal>(goal, 1e-3);
  prob.timeout = 5.0;
  prob.seed = seed;
  return prob;
}

} // namespace

TEST(PrmPlanner, NullArgsThrow) {
  auto f = make_fixture(false);
  EXPECT_THROW(make_prm_planner(params(1), nullptr, f.scene), std::runtime_error);
  EXPECT_THROW(make_prm_planner(params(1), f.robot, nullptr), std::runtime_error);
}

TEST(PrmPlanner, RegistryStubFailsLoudly) {
  auto f = make_fixture(false);
  const auto ids = registered_planners();
  EXPECT_NE(std::find(ids.begin(), ids.end(), "prm"), ids.end());
  PlannerParams pp;
  pp.algorithm = "prm";
  EXPECT_THROW(make_planner(pp, f.robot, f.scene), std::runtime_error);
}

TEST(PrmPlanner, BuildStatsPopulated) {
  auto f = make_fixture(true);
  PrmBuildStats stats;
  const auto prm = make_prm_planner(params(7), f.robot, f.scene, &stats);
  EXPECT_GT(stats.nodes, 100u); // most of 800 samples are free in this cell
  EXPECT_EQ(stats.node_candidates, 800u);
  EXPECT_GT(stats.edges, stats.nodes); // k-nearest connectivity ⇒ more edges than nodes
  EXPECT_GT(stats.collision_configs, 0u);
  EXPECT_GT(stats.build_seconds, 0.0);
}

TEST(PrmPlanner, SolvesAroundWall) {
  auto f = make_fixture(true);
  const auto prm = make_prm_planner(params(3), f.robot, f.scene);
  const PlanningResult r = prm->plan(problem(q2(-1, -1), q2(1, -1), 42));
  ASSERT_EQ(r.status, PlanningStatus::Success) << r.message;
  EXPECT_TRUE(path_is_free(f, r.path, 0.02));
  EXPECT_LT((r.path.front() - q2(-1, -1)).norm(), 1e-6);
  EXPECT_LT((r.path.back() - q2(1, -1)).norm(), 1e-6);
  // Routing around the wall means the path must rise above the padded top (y > 0.4 somewhere).
  const double max_y =
      std::max_element(r.path.begin(), r.path.end(),
                       [](const JointPosition &a, const JointPosition &b) { return a.y() < b.y(); })
          ->y();
  EXPECT_GT(max_y, 0.4);
}

TEST(PrmPlanner, OneRoadmapAnswersManyQueries) {
  auto f = make_fixture(true);
  const auto prm = make_prm_planner(params(5), f.robot, f.scene);
  for (const auto &[s, g] : std::vector<std::pair<JointPosition, JointPosition>>{
           {q2(-1, -1), q2(1, -1)}, {q2(-1.5, 0), q2(1.5, 0)}, {q2(-1, 1), q2(1, -1.5)}}) {
    const PlanningResult r = prm->plan(problem(s, g, 1));
    ASSERT_EQ(r.status, PlanningStatus::Success) << r.message;
    EXPECT_TRUE(path_is_free(f, r.path, 0.02));
  }
}

TEST(PrmPlanner, DeterministicPerSeed) {
  auto f = make_fixture(true);
  const auto a = make_prm_planner(params(9), f.robot, f.scene);
  const auto b = make_prm_planner(params(9), f.robot, f.scene); // same build seed ⇒ same roadmap
  const PlanningResult ra = a->plan(problem(q2(-1, -1), q2(1, -1), 100));
  const PlanningResult rb = b->plan(problem(q2(-1, -1), q2(1, -1), 100));
  ASSERT_EQ(ra.status, PlanningStatus::Success);
  ASSERT_EQ(ra.path.size(), rb.path.size());
  for (std::size_t i = 0; i < ra.path.size(); ++i) {
    EXPECT_LT((ra.path[i] - rb.path[i]).norm(), 1e-12) << "waypoint " << i << " differs";
  }
}

File: bindings/python/tests/test_prm.py
"""Roadmap R5 — PrmPlanner bindings: build a roadmap once (PrmBuildStats), answer many queries
around a wall, determinism per seed, and the 'prm' registry stub failing loudly."""

import numpy as np
import pytest

import quevedomp as q

GANTRY_2D = """<robot name="gantry2d">
  <link name="base"/>
  <joint name="jx" type="prismatic"><parent link="base"/><child link="cx"/>
    <origin xyz="0 0 0"/><axis xyz="1 0 0"/><limit lower="-2" upper="2" effort="10" velocity="1"/></joint>
  <link name="cx"/>
  <joint name="jy" type="prismatic"><parent link="cx"/><child link="ee"/>
    <origin xyz="0 0 0"/><axis xyz="0 1 0"/><limit lower="-2" upper="2" effort="10" velocity="1"/></joint>
  <link name="ee"><collision><geometry><sphere radius="0.1"/></geometry></collision></link>
</robot>"""


@pytest.fixture(scope="module")
def cell():
    model = q.RobotModel.from_urdf(GANTRY_2D)
    robot = q.RobotInstance(model)
    env = q.SceneDescription()
    env.add("wall", q.BoxShape(np.array([0.1, 1.25, 0.5])),
            q.Transform.from_translation(np.array([0.0, -0.75, 0.0])))
    scene = q.make_static_scene(model, env)
    return model, robot, scene


def _params(seed):
    p = q.PrmParams()
    p.num_nodes = 800
    p.k_neighbors = 12
    p.edge_resolution = 0.02
    p.seed = seed
    return p


def _problem(start, goal, seed):
    prob = q.PlanningProblem()
    prob.start = np.asarray(start, dtype=float)
    prob.goal = q.JointGoal(np.asarray(goal, dtype=float), 1e-3)
    prob.timeout = 5.0
    prob.seed = seed
    return prob


def _free(scene, robot, path, res=0.02):
    ws = scene.make_workspace()
    for a, b in zip(path[:-1], path[1:]):
        if not q.check_edge(scene, robot, np.asarray(a), np.asarray(b), res, q.QueryOptions(), ws).valid:
            return False
    return True


def test_registry_stub_fails_loudly(cell):
    model, robot, scene = cell
    assert "prm" in q.registered_planners()
    pp = q.PlannerParams()
    pp.algorithm = "prm"
    with pytest.raises(Exception):
        q.make_planner(pp, robot, scene)


def test_build_and_solve_around_wall(cell):
    model, robot, scene = cell
    prm, stats = q.make_prm_planner(_params(3), robot, scene)
    assert stats.nodes > 100
    assert stats.edges > stats.nodes
    assert stats.build_seconds > 0.0

    r = prm.plan(_problem([-1.0, -1.0], [1.0, -1.0], 42))
    assert r.ok(), r.message
    assert _free(scene, robot, list(r.path))
    assert max(float(w[1]) for w in r.path) > 0.4  # routes up and over the wall


def test_one_roadmap_many_queries(cell):
    model, robot, scene = cell
    prm, _ = q.make_prm_planner(_params(5), robot, scene)
    for s, g in [([-1.0, -1.0], [1.0, -1.0]), ([-1.5, 0.0], [1.5, 0.0])]:
        r = prm.plan(_problem(s, g, 1))
        assert r.ok(), r.message
        assert _free(scene, robot, list(r.path))


@pytest.fixture(scope="module")
def divided_cell():
    """The gantry with a FULL-WIDTH divider at y=0: the free space splits into a lower (y<0) and
    an upper (y>0) strip with no collision-free crossing — the roadmap must split into components."""
    model = q.RobotModel.from_urdf(GANTRY_2D)
    robot = q.RobotInstance(model)
    env = q.SceneDescription()
    env.add("divider", q.BoxShape(np.array([2.5, 0.15, 0.5])),
            q.Transform.from_translation(np.array([0.0, 0.0, 0.0])))
    scene = q.make_static_scene(model, env)
    return model, robot, scene


def test_split_roadmap_diagnostics(divided_cell):
    model, robot, scene = divided_cell
    p = _params(7)
    p.num_nodes = 600
    p.export_roadmap = True
    prm, stats = q.make_prm_planner(p, robot, scene)

    # A full-width divider disconnects the free space, so the roadmap is not one component.
    assert stats.num_components >= 2
    assert 0 < stats.largest_component <= stats.nodes

    # Geometry export is self-consistent (studio draws this).
    assert len(stats.roadmap_nodes) == stats.nodes
    assert len(stats.roadmap_component) == stats.nodes
    assert all(0 <= c < stats.num_components for c in stats.roadmap_component)
    for a, b in stats.roadmap_edges:
        assert 0 <= a < stats.nodes and 0 <= b < stats.nodes

    # A query across the divider fails with a diagnostic, not a bare "no path".
    r = prm.plan(_problem([0.0, -1.0], [0.0, 1.0], 1))
    assert not r.ok()
    assert "DISJOINT" in r.message or "narrow passage" in r.message, r.message


def test_export_roadmap_off_by_default(cell):
    model, robot, scene = cell
    _, stats = q.make_prm_planner(_params(4), robot, scene)  # export_roadmap defaults False
    assert stats.num_components >= 1          # counts always come back
    assert len(stats.roadmap_nodes) == 0      # geometry only on request
    assert len(stats.roadmap_edges) == 0


def test_deterministic_per_seed(cell):
    model, robot, scene = cell
    a, _ = q.make_prm_planner(_params(9), robot, scene)
    b, _ = q.make_prm_planner(_params(9), robot, scene)
    ra = a.plan(_problem([-1.0, -1.0], [1.0, -1.0], 100))
    rb = b.plan(_problem([-1.0, -1.0], [1.0, -1.0], 100))
    assert ra.ok() and len(ra.path) == len(rb.path)
    for pa, pb in zip(ra.path, rb.path):
        assert np.allclose(pa, pb, atol=1e-12)
